javascript/ql/lib/semmle/javascript/Actions.qll

```diff
@@ -0,0 +1,322 @@
+    string repositoryOwner;
+    /** The name of the repository where the Action comes from, e.g. `checkout` in `actions/checkout@v2`. */
+    string repositoryName;
+    /** The version reference used when checking out the Action, e.g. `v2` in `actions/checkout@v2`. */
+    string version;
+
+    Uses() {
+      step.lookup("uses") = this and
+      // Simple regular expression to split up an Action reference `owner/repo@version` into its components.
+      exists(string regexp | regexp = "([^/]+)/([^/@]+)@(.+)" |
+        repositoryOwner = this.getValue().regexpCapture(regexp, 1) and
+        repositoryName = this.getValue().regexpCapture(regexp, 2) and
+        version = this.getValue().regexpCapture(regexp, 3)
+      )
+    }
+
+    /** Gets the step this field belongs to. */
+    Step getStep() { result = step }
+
+    /** Gets the owner and name of the repository where the Action comes from, e.g. `actions/checkout` in `actions/checkout@v2`. */
+    string getGitHubRepository() { result = repositoryOwner + "/" + repositoryName }
+
+    /** Gets the version reference used when checking out the Action, e.g. `v2` in `actions/checkout@v2`. */
+    string getVersion() { result = version }
+  }
+
+  /**
+   * A `with` field within an Actions job step, which references an action as a reusable unit of code.
+   * See https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepswith.
+   *
+   * For example:
+   * ```
+   * with:
+   *   arg1: 1
+   *   arg2: abc
+   * ```
+   */
+  class With extends YAMLNode, YAMLMapping {
+    Step step;
+
+    With() { step.lookup("with") = this }
+
+    /** Gets the step this field belongs to. */
+    Step getStep() { result = step }
+  }
+
+  /**
+   * A `ref:` field within an Actions `with:` specific to `actions/checkout` action.
+   *
+   * For example:
+   * ```
+   * uses: actions/checkout@v2
+   * with:
+   *   ref: ${{ github.event.pull_request.head.sha }}
+   * ```
+   */
+  class Ref extends YAMLNode, YAMLString {
+    With with;
+
+    Ref() { with.lookup("ref") = this }
+
+    /** Gets the `with` field this field belongs to. */
+    With getWith() { result = with }
+  }
+
+  /**
+   * A `run` field within an Actions job step, which runs command-line programs using an operating system shell.
+   * See https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsrun.
+   */
+  class Run extends YAMLNode, YAMLString {
+    Step step;
+
+    Run() { step.lookup("run") = this }
+
+    /** Gets the step that executes this `run` command. */
+    Step getStep() { result = step }
+
+    /**
+     * Holds if `${{ e }}` is a GitHub Actions expression evaluated within this `run` command.
+     * See https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions.
+     */
+    string getAReferencedExpression() {
```

Comment:
Could we avoid the TODO below if we rename the predicate to be more explicit about the limitations, i.e. `getASimpleReferenceExpression`?

In any case, the limitation should also have a positive and a negative example since the regexps are hard to parse mentally.